pkg/modules/spanpercentile/implspanpercentile/translator.go

```diff
@@ -0,0 +1,208 @@
+package implspanpercentile
+
+import (
+	"fmt"
+	"github.com/SigNoz/signoz/pkg/querybuilder"
+	"github.com/SigNoz/signoz/pkg/querybuilder/resourcefilter"
+	"github.com/SigNoz/signoz/pkg/telemetrytraces"
+	qbtypes "github.com/SigNoz/signoz/pkg/types/querybuildertypes/querybuildertypesv5"
+	"github.com/SigNoz/signoz/pkg/types/spanpercentiletypes"
+	"github.com/huandu/go-sqlbuilder"
+	"strings"
+)
+
+func buildSpanPercentileQuery(req *spanpercentiletypes.SpanPercentileRequest) (*qbtypes.QueryRangeRequest, error) {
+	if err := req.Validate(); err != nil {
+		return nil, err
+	}
+
+	builder := &spanPercentileCTEBuilder{
+		start:   querybuilder.ToNanoSecs(req.Start),
+		end:     querybuilder.ToNanoSecs(req.End),
+		spanID:  req.SpanID,
+		filter:  req.Filter,
+		request: req,
+	}
+
+	stmt := builder.build()
+
+	query := qbtypes.QueryEnvelope{
+		Type: qbtypes.QueryTypeClickHouseSQL,
+		Spec: qbtypes.ClickHouseQuery{
+			Name:  "span_percentile",
+			Query: stmt.Query,
+		},
+	}
+
+	return &qbtypes.QueryRangeRequest{
+		SchemaVersion: "v5",
+		Start:         req.Start,
+		End:           req.End,
+		RequestType:   qbtypes.RequestTypeScalar,
+		CompositeQuery: qbtypes.CompositeQuery{
+			Queries: []qbtypes.QueryEnvelope{query},
+		},
+		FormatOptions: &qbtypes.FormatOptions{
+			FormatTableResultForUI: true,
+		},
+	}, nil
+}
+
+type cteNode struct {
+	name string
+	sql  string
+	args []any
+}
+
+type spanPercentileCTEBuilder struct {
+	start   uint64
+	end     uint64
+	spanID  string
+	filter  *qbtypes.Filter
+	request *spanpercentiletypes.SpanPercentileRequest
+	ctes    []cteNode
+}
+
+func (b *spanPercentileCTEBuilder) build() *qbtypes.Statement {
+	b.buildResourceFilterCTE()
+	b.buildBaseSpansCTE()
+	b.buildTargetSpanCTE()
+
+	mainSQL, mainArgs := b.buildMainQuery()
+
+	var cteFragments []string
+	var cteArgs [][]any
+
+	for _, cte := range b.ctes {
+		cteFragments = append(cteFragments, fmt.Sprintf("%s AS (%s)", cte.name, cte.sql))
+		cteArgs = append(cteArgs, cte.args)
+	}
+
+	finalSQL := querybuilder.CombineCTEs(cteFragments) + mainSQL
+	finalArgs := querybuilder.PrependArgs(cteArgs, mainArgs)
+
+	interpolatedSQL := b.interpolateArgs(finalSQL, finalArgs)
+
+	return &qbtypes.Statement{
+		Query: interpolatedSQL,
+		Args:  nil,
+	}
+}
+
+func (b *spanPercentileCTEBuilder) buildResourceFilterCTE() {
+	if b.filter == nil || b.filter.Expression == "" {
+		return
+	}
+
+	startBucket := b.start/querybuilder.NsToSeconds - querybuilder.BucketAdjustment
+	endBucket := b.end / querybuilder.NsToSeconds
+
+	sb := sqlbuilder.NewSelectBuilder()
+	sb.Select("fingerprint")
+	sb.From(fmt.Sprintf("%s.%s", resourcefilter.TracesDBName, resourcefilter.TraceResourceV3TableName))
+	sb.Where(
+		sb.GE("seen_at_ts_bucket_start", fmt.Sprintf("%d", startBucket)),
+		sb.LE("seen_at_ts_bucket_start", fmt.Sprintf("%d", endBucket)),
+	)
+	sb.Where(fmt.Sprintf("(%s)", b.filter.Expression))
```

Comment:
can you directly add filter to a where clause ? we do a lot of metadata update and traformation before building a where clause from expression